ocaml/typing/typecore.ml

```diff
@@ -374,8 +374,19 @@ type expected_mode =
     field and [mode]: this field being [true] while [mode] being [global] is
     sensible, but not very useful as it will fail all expressions. *)
 
-    tuple_modes : Value.r list;
-    (** For t in tuple_modes, t <= regional_to_global mode *)
+    tuple_modes : Value.r list option;
+    (** No invariant between this and [mode]. It is UNSOUND to ignore this field.
+        If this is [Some [x0; x1; ..]]:
+
+        - If we are type checking [Pexp_tuple [e0; e1; ..]], then we check
+          [e0 <= x0] and [e1 <= x1] and so on. Moreover, we check
+          [regional_to_local(join(e0, e1, ..)) <= mode].
+
+        - If we are type checking "non-concrete" [Pexp_constraint] etc, this
+          field must be preserved.
+
+        - If we are type checking other expression [e], then we check [e <= x0]
+          and [e <= x1] and so on. Moreover, [e <= mode]. *)
```

Comment:
`Moreover` should be `and we also check` or similar. (`Moreover` is ambiguous between "we also need to check e <= mode" and "that already implies e <= mode")